app/src/heartbeat/heartbeat_storage_desktop.cc

```diff
@@ -0,0 +1,125 @@
+const char kHeartbeatFilename[] = "HEARTBEAT_STORAGE";
+
+// Returns the mutex that protects accesses to the storage file.
+Mutex& FileMutex() {
+  static Mutex* mutex_ = new Mutex(Mutex::kModeNonRecursive);
+  return *mutex_;
+}
+
+std::string GetFilename(std::string& error) {
+  std::string app_dir =
+      AppDataDir(kHeartbeatDir, /*should_create=*/true, &error);
+  if (app_dir.empty()) {
+    return "";
+  }
+
+  // TODO(almostmatt): does this need to be different between windows/mac/linux?
+  return app_dir + "/" + kHeartbeatFilename;
+}
+
+}  // namespace
+
+HeartbeatStorageDesktop::HeartbeatStorageDesktop()
+    : filename_(GetFilename(error_)) {
+  MutexLock lock(FileMutex());
+
+  // Ensure the file exists, otherwise the first attempt to read it would
+  // fail.
+  std::ofstream file(filename_, std::ios_base::app);
+
+  if (!file) {
+    error_ = "Unable to open '" + filename_ + "'.";
+  }
+}
+
+const LoggedHeartbeats* HeartbeatStorageDesktop::Read() {
+  MutexLock lock(FileMutex());
+  error_ = "";
+
+  // Open the file and seek to the end
+  std::ifstream file(filename_, std::ios_base::binary | std::ios_base::ate);
+  if (!file) {
+    // TODO(almostmatt): if file does not exist, maybe return a default
+    // instance.
+    error_ = "Unable to open '" + filename_ + "' for reading.";
+    return nullptr;
+  }
+  // Current position in file is buffer size.
+  std::streamsize buffer_len = file.tellg();
+  // Rewind to start of the file, then read into a buffer.
+  file.seekg(0, std::ios_base::beg);
+  char buffer[buffer_len];
+  file.read(buffer, buffer_len);
+  // Verify that the buffer is a valid flatbuffer.
+  ::flatbuffers::Verifier verifier(reinterpret_cast<const uint8_t*>(buffer),
+                                   buffer_len);
+  if (!VerifyLoggedHeartbeatsBuffer(verifier)) {
+    // TODO(almostmatt): maybe delete the file if it contains corrupted data
+    error_ =
+        "Failed to parse contents of " + filename_ + " as LoggedHeartbeats.";
+    return nullptr;
+  }
+  return GetLoggedHeartbeats(&buffer);
+}
+
+bool HeartbeatStorageDesktop::Write(const LoggedHeartbeats* heartbeats) const {
+  MutexLock lock(FileMutex());
+  error_ = "";
+
+  // Clear the file before writing.
+  std::ofstream file(filename_, std::ios_base::trunc | std::ios_base::binary);
+  if (!file) {
+    error_ = "Unable to open '" + filename_ + "' for writing.";
+    return false;
+  }
+
+  // LoggedHeartbeats is just part of a flatbuffer. To get the entire buffer,
+  // Create a new flat buffer builder based on the LoggedHeartbeats instance.
+  // TODO(almostmatt): This does not seem good. Find a better way to write.
+  flatbuffers::FlatBufferBuilder fbb(1024);
+  auto heartbeats_unique_ptr = UnPackLoggedHeartbeats(heartbeats);
+  auto heartbeats_offset =
+      CreateLoggedHeartbeats(fbb, heartbeats_unique_ptr.get());
+  fbb.Finish(heartbeats_offset);
+  file.write((char*)fbb.GetBufferPointer(), fbb.GetSize());
```

Comment:
The logic here is not good. All of the documentation and examples I could find for writing a flatbuffer to a disk had a pointer to the buffer of bytes. But I could not find a good way to go from the LoggedHeartbeats object to the buffer pointer, only the flat buffer builder has a convenient getter for the buffer pointer.

If there is a more convenient way to get the buffer pointer from the class I would be happy to use that.

Reply:
@jonsimantov  Are you familiar about how this is done with Flatbuffer?

Reply:
I spoke a bit with Jon about flatbuffers and serializations. Seems that it is correct that the buffer pointer and size is needed to serialize, and LoggedHeartbeats is just a part of the buffer, not the whole thing.

Since LoggedHeartbeats is not too easy to modify or write, and a byte array is too implementation-specific, I am considering changing these methods to input and output a simple C struct that contains a last-logged field and a std::map from user agent to a list of date strings. This will mean a little bit of extra code to convert between the two, but the controller code would no longer have to deal with flatbuffer builders to modify the in-memory object.
In that case, the benefit of flatbuffers for serialization is mainly just that it allows us to verify the validity of the structure when it reads a byte buffer from a file.

Reply:
It is good idea to free controller from handling flatbuffer.

Given that there is not change in this file, I assume your change will come in in the future?